lua40a/doc/lua.man
.\" lua.man,v 1.2 2000/04/18 00:21:27 lhf Exp
.TH LUA 1 "2000/04/18 00:21:27"
.SH NAME
lua \- Lua interpreter
.SH SYNOPSIS
.B lua
[
.I arguments
]
.SH DESCRIPTION
.B lua
is the stand-alone Lua interpreter.
It loads and executes Lua programs,
either in textual source form or
in precompiled binary form,
as output by
.BR luac ,
the Lua compiler.
.B lua
can be used as a batch interpreter and also interactively.
.LP
The
.I arguments
can be options, assignments, or filenames,
and are processed in order,
from left to right.
.LP
Options start with
.B \-
and are described below.
.LP
An assignment is an argument of the form
.BR a=b ,
which assigns the string
.RB ` b '
to the global variable
.BR a .
Note that no quotes are needed around the string if it does not contain spaces
or other characters special to the shell.
This is for convenience only.
(In general,
you should be careful when using quotes and spaces on the command line
because they are usually handled by the shell.)
.LP
If the argument is neither an option nor an assignment,
then it is assumed to be a filename,
which is then loaded and executed.
.LP
If no arguments are given,
then
.B "\-v \-i"
is assumed when the standard input is a terminal;
otherwise,
.B \-
is assumed.
.SH OPTIONS
.TP
.B \-
load the standard input as a file,
that is,
not interactively,
even when the standard input is a terminal.
.TP
.B \-d
turn debugging on.
Individual chunks may
still control the generation of debug information with
.B $debug
and
.BR $nodebug .
If debugging is on, then additional information is printed when an error occurs.
.TP
.BI \-e " stat"
execute statement
.IR stat .
You will need to quote
.I stat 
if it contains spaces or quotes.
.TP
.BI \-f " file"
collect all remaining arguments as strings into a global table named
.B arg
and then execute
.IR file .
The arguments in
.B arg
start at 0,
which contains
.RI ` file '.
The index of the last argument is stored in
.BR "arg.n" .
.TP
.B \-i
enter interactive mode,
displaying a prompt.
In this mode,
.B lua
reads lines from the standard input and executes them as they are read.
Each line must contain a complete statement.
To span a statement across several lines, end each line with a backslash
.BR `\e' .
The prompt shown is the value of the global variable
.BR _PROMPT ,
if this value is a string.
So,
to change the prompt,
set
.B _PROMPT
to a string of your choice.
You can do that after calling the interpreter
or on the command line with
.BR "_PROMPT=\'lua: \'" ,
for example.
(Note the need for quotes, because the string contains a space.)
The default prompt is ``> ''.
.TP
.B \-q
enter interactive mode,
but without displaying a prompt.
.TP
.BI \-s n
set the stack size to
.IB n .
If present,
this must be the first option.
Note that
.I n
is in the same argument as
.BR -s .
For example,
to specify a stack size of 2000,
use
.BR -s2000 .
.TP
.B \-v
print version information.
.SH "SEE ALSO"
.BR luac (1)
.br
http://www.tecgraf.puc-rio.br/lua/
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH AUTHORS
R. Ierusalimschy,
L. H. de Figueiredo,
and
W. Celes
(lua@tecgraf.puc-rio.br)
.\" EOF


lua31/history
This is version 3.1 (alpha).

* Changes from version 3.0 to 3.1 (alpha)
  + NEW FEATURE: anonymous functions with closures (via "upvalues").
  + new syntax:
    - local variables in chunks.
    - better scope control with DO block END.
    - constructors can now be also written: { record-part; list-part }.
    - more general syntax for function calls and lvalues, e.g.:
      f().x=1
      o:f(x,y):g(z)
  + major code re-organization and clean-up; reduced module interdependecies.
  + support for multiple contexts.
  + new traversal functions "foreach" and "foreachvar".
  + changing lua to use doubles instead of floats is now trivial.
  + complete debug information stored in pre-compiled chunks.

* Changes from version 2.5 to 3.0
  + NEW CONCEPT: "tag methods".
    Tag methods replace fallbacks as the meta-mechanism for extending the
    semantics of Lua. Whereas fallbacks had a global nature, tag methods
    work on objects having the same tag (e.g., groups of tables).
    Existing code that uses fallbacks should work without change.
  + new, general syntax for constructors {[exp] = exp, ... }.
  + support for handling variable number of arguments in functions (varargs).
  + support for conditional compilation ($if ... $else ... $end).
  + cleaner semantics in API simplifies host code.
  + better support for writing libraries (auxlib.h).
  + better type checking and error messages in the standard library.
  + luac can now also undump.

* Changes from version 2.4 to 2.5
  + io and string libraries are now based on pattern matching;
    the old libraries are still available for compatibility
  + dofile and dostring can now return values (via return statement)
  + better support for 16- and 64-bit machines
  + expanded documentation, with more examples

* Changes from version 2.2 to 2.4
  + external compiler creates portable binary files that can be loaded faster
  + interface for debugging and profiling
  + new "getglobal" fallback
  + new functions for handling references to Lua objects
  + new functions in standard lib
  + only one copy of each string is stored
  + expanded documentation, with more examples

* Changes from version 2.1 to 2.2
  + functions now may be declared with any "lvalue" as a name
  + garbage collection of functions
  + support for pipes

* Changes from version 1.1 to 2.1
  + object-oriented support
  + fallbacks
  + simplified syntax for tables
  + many internal improvements


lua40a/src/lua/lua.c
/*
** $Id: lua.c,v 1.1 2000/07/20 10:16:44 jeske Exp $
** Lua stand-alone interpreter
** See Copyright Notice in lua.h
*/


#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "lua.h"
#include "luadebug.h"
#include "lualib.h"


#ifndef PROMPT
#define PROMPT		"> "
#endif

#ifdef _POSIX_SOURCE
#include <unistd.h>
#else
static int isatty (int x) { return x==0; }  /* assume stdin is a tty */
#endif

typedef void (*handler)(int);  /* type for signal actions */

static void laction (int i);


static lua_Hook old_linehook = NULL;
static lua_Hook old_callhook = NULL;


static handler lreset (void) {
  return signal(SIGINT, laction);
}


static void lstop (void) {
  lua_setlinehook(lua_state, old_linehook);
  lua_setcallhook(lua_state, old_callhook);
  lreset();
  lua_error("interrupted!");
}


static void laction (int i) {
  (void)i;  /* to avoid warnings */
  signal(SIGINT, SIG_DFL); /* if another SIGINT happens before lstop,
                              terminate process (default action) */
  old_linehook = lua_setlinehook(lua_state, (lua_Hook)lstop);
  old_callhook = lua_setcallhook(lua_state, (lua_Hook)lstop);
}


static int ldo (int (*f)(lua_State *L, const char *), const char *name) {
  int res;
  handler h = lreset();
  res = f(lua_state, name);  /* dostring | dofile */
  signal(SIGINT, h);  /* restore old action */
  return res;
}


static void print_message (void) {
  fprintf(stderr,
  "usage: lua [options].  Available options are:\n"
  "  -        execute stdin as a file\n"
  "  -d       turn debug on\n"
  "  -e stat  execute string `stat'\n"
  "  -f name  execute file `name' with remaining arguments in table `arg'\n"
  "  -i       enter interactive mode with prompt\n"
  "  -q       enter interactive mode without prompt\n"
  "  -sNUM    set stack size to NUM (must be first option)\n"
  "  -v       print version information\n"
  "  a=b      set global `a' to string `b'\n"
  "  name     execute file `name'\n"
);
}


static void print_version (void) {
  printf("%s  %s\n", LUA_VERSION, LUA_COPYRIGHT);
}


static void assign (char *arg) {
  char *eq = strchr(arg, '=');
  *eq = '\0';  /* spilt `arg' in two strings (name & value) */
  lua_pushstring(eq+1);
  lua_setglobal(arg);
}


static lua_Object getargs (char *argv[]) {
  lua_Object args = lua_createtable();
  int i;
  for (i=0; argv[i]; i++) {
    /* arg[i] = argv[i] */
    lua_pushobject(args); lua_pushnumber(i);
    lua_pushstring(argv[i]); lua_settable();
  }
  /* arg.n = maximum index in table `arg' */
  lua_pushobject(args); lua_pushstring("n");
  lua_pushnumber(i-1); lua_settable();
  return args;
}


static void l_getargs (void) {
  char **argv = (char **)lua_getuserdata(lua_getparam(1));
  lua_pushobject(getargs(argv));
}


static void file_input (const char *argv) {
  int result = ldo(lua_dofile, argv);
  if (result) {
    if (result == 2) {
      fprintf(stderr, "lua: cannot execute file ");
      perror(argv);
    }
    exit(1);
  }
}


static void manual_input (int version, int prompt) {
  int cont = 1;
  if (version) print_version();
  while (cont) {
    char buffer[BUFSIZ];
    int i = 0;
    lua_beginblock();
    if (prompt) {
      const char *s = lua_getstring(lua_getglobal("_PROMPT"));
      if (!s) s = PROMPT;
      fputs(s, stdout);
    }
    for(;;) {
      int c = getchar();
      if (c == EOF) {
        cont = 0;
        break;
      }
      else if (c == '\n') {
        if (i>0 && buffer[i-1] == '\\')
          buffer[i-1] = '\n';
        else break;
      }
      else if (i >= BUFSIZ-1) {
        fprintf(stderr, "lua: input line too long\n");
        break;
      }
      else buffer[i++] = (char)c;
    }
    buffer[i] = '\0';
    ldo(lua_dostring, buffer);
    lua_endblock();
  }
  printf("\n");
}


int main (int argc, char *argv[]) {
  int i = 1;
  if (i < argc && argv[1][0] == '-' && argv[1][1] == 's') {
    int stacksize = atoi(&argv[1][2]);
    if (stacksize == 0) {
      fprintf(stderr, "lua: invalid stack size ('%s')\n", &argv[1][2]);
      exit(1);
    }
    i++;
    lua_state = lua_newstate("stack", stacksize, NULL);
  }
  else
    lua_state = lua_newstate(NULL);
  lua_userinit();
  lua_pushuserdata(argv);
  lua_pushcclosure(l_getargs, 1);
  lua_setglobal("getargs");
  if (i >= argc) {  /* no other arguments? */
    if (isatty(0)) {
      manual_input(1, 1);
    }
    else
      ldo(lua_dofile, NULL);  /* executes stdin as a file */
  }
  else for (; i<argc; i++) {
    if (argv[i][0] == '-') {  /* option? */
      switch (argv[i][1]) {
        case 0:
          ldo(lua_dofile, NULL);  /* executes stdin as a file */
          break;
        case 'i':
          manual_input(0, 1);
          break;
        case 'q':
          manual_input(0, 0);
          break;
        case 'd':
          lua_setdebug(lua_state, 1);
          if (i+1 >= argc) {  /* last argument? */
            manual_input(1, 1);
          }
          break;
        case 'v':
          print_version();
          break;
        case 'e':
          i++;
          if (i >= argc) {
            print_message();
            exit(1);
          }
          if (ldo(lua_dostring, argv[i]) != 0) {
            fprintf(stderr, "lua: error running argument `%s'\n", argv[i]);
            exit(1);
          }
          break;
        case 'f':
          i++;
          if (i >= argc) {
            print_message();
            exit(1);
          }
          lua_pushobject(getargs(argv+i));  /* collect remaining arguments */
          lua_setglobal("arg");
          file_input(argv[i]);
          goto endloop;  /* stop scanning arguments */
          break;
        case 's':
          fprintf(stderr, "lua: stack size (`-s') must be the first option\n");
          exit(1);
        default:
          print_message();
          exit(1);
      }
    }
    else if (strchr(argv[i], '='))
      assign(argv[i]);
    else
      file_input(argv[i]);
  }
  endloop:
#ifdef DEBUG
  lua_close();
#endif
  return 0;
}

